### PR TITLE
Fix "snapdev" property issues

### PR DESCRIPTION
**Pull request by loli10K**
### Description

When inheriting the 'snapdev' property to "visible" we don't call `zfs_prop_set_special()`: this prevents device nodes from being created. Because 'snapdev' is the only "special" property that is also inheritable we need to call `zfs_prop_set_special()` even when we're not reverting to the received value (`zfs inherit -S`).

Additionally fix a NULL pointer dereference bug triggered when setting the "snapdev=hidden" property twice.

References: https://github.com/zfsonlinux/zfs/pull/6131

### Motivation and Context

Fix https://github.com/zfsonlinux/zfs/issues/6175

### How Has This Been Tested?

Test case added to zvol_misc

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by loli10K** [comment]
@bprotopopov can you please review the proposed changes? We could also try to squeeze in your idea to optimize minors creation/removal when setting the `snapdev` property if the value does not change, but maybe in another pull request?

**Comment by bprotopopov** [comment]
Sure I'll take a look

Typos courtesy of my iPhone

On May 27, 2017, at 5:11 AM, LOLi <notifications@github.com<mailto:notifications@github.com>> wrote:


@bprotopopov<https://github.com/bprotopopov> can you please review the proposed changes? We could also try to squeeze in your idea to optimize minors creation/removal when setting the snapdev property if the value does not change, but maybe in another pull request?

**Comment by bprotopopov** [comment]
@loli10K 
I can do the optimization of minor/major creation when addressing issue #6175 (which I will rename if needed to further clarify the scope). How soon do you need the review of this PR?

**Comment by loli10K** [comment]
> when addressing issue #6175 

This PR should already address that issue, as stated in the "Motivation and Context" section.

> How soon do you need the review of this PR?

I have no deadlines to meet, whenever you can is fine

File: module/zfs/zvol.c [hunk 1]
```diff
@@ -2010,7 +2010,7 @@ zvol_remove_minors_impl(const char *name)
 static void
 zvol_remove_minor_impl(const char *name)
 {
-	zvol_state_t *zv, *zv_next;
+	zvol_state_t *zv = NULL, *zv_next;
```
Comment (loli10K) [inline module/zfs/zvol.c]: `zv` is always initialized at 2030 with the value of `list_head(&zvol_state_list)` and yes, `zv_next` might not be initialized if the list is empty, in which case `zv_next` is not used.

Comment (bprotopopov) [inline module/zfs/zvol.c]: It is a matter of style, I suppose, I would not initialize the zv explicitly then.

**Comment by bprotopopov** [comment]
@loli10K 

I would rather fix the NULL pointer dereference separately from this RP as it seems to address a separate set of issues related to inheritance, etc. Can you submit a separate PR ? Or you can delegate it to me, no problem.

I am reviewing the inheritance-related changes and will comment more inline.

File: module/zfs/zfs_ioctl.c [hunk 2]
```diff
@@ -2819,8 +2777,49 @@ zfs_ioc_inherit_prop(zfs_cmd_t *zc)
...
+		goto errout;
+	}
+
+	pair = nvlist_next_nvpair(dummy, NULL);
+	if (pair == NULL) {
+		err = SET_ERROR(EINVAL);
+	} else {
+		err = zfs_prop_set_special(zc->zc_name, source, pair);
+		if (err == -1) /* property is not "special", needs handling */
+			err = dsl_prop_inherit(zc->zc_name, zc->zc_value,
```
Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: I am looking at `zfs_ioc_inherit_prop()`, and it seems that the refactoring has changed the logic, e.g. the function now calls `zfs_set_prop_special()` when `received == FALSE`, which it did not before. Is this intended ? And if so, could you please add comments as to the intent ?

**Comment by loli10K** [comment]
> I would rather fix the NULL pointer dereference separately from this RP as it seems to address a separate set of issues related to inheritance, etc. Can you submit a separate PR ? Or you can delegate it to me, no problem.

The contributing guidelines state "_Please attempt to limit pull requests to a single commit which resolves one specific issue_" which i understand and support completely: it makes backporting and bisecting easier and helps the review process. On the other hand pushing many small pull requests can generate a lot of noise, overhead and place a burden on reviewers, and i'd like to avoid that.

Fixing the NULL ptr deref won't probably require any backporting since it was introduced only recently, and it's also needed by the other "snapdev" fixes in this PR, so it seemed only natural to me to squeeze it in.

If you have other reasons and still want to submit the fix as separate PR i'll drop that change here and rebase on master once yours gets merged.

File: module/zfs/zfs_ioctl.c (hunk 2 above)
Comment (loli10K) [inline module/zfs/zfs_ioctl.c]: Yes this is intended, the commit message should explain it: 

> When inheriting the "snapdev" property to we don't always call zfs_prop_set_special(): this prevents device nodes from being created in certain situations.

> Because "snapdev" is the only *special* property that is also inheritable we need to call zfs_prop_set_special() even when we're not reverting it to the received value

**Comment by bprotopopov** [comment]
@loli10K 
I think we might need to iterate a bit longer over your changes in this PR, so I thought a simpler portion of this PR (the one-liner with NULL pointer fix) could be pushed quickly to address this issue on the master, while the rest of this PR is being reviewed ? That's the only reason, really.

File: module/zfs/zfs_ioctl.c (hunk 2 above)
Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: @loli10K 
it seems that the change is not limited to the snapdev property. If you intended to affect snapdev only, would it make sense to set received == TRUE unconditionally in the user-space for the snapdev only ? If you did not intend to limit the effects to snapdev, you should probably indicate why this is OK, and perhaps revise the commit message ?

Also, this type of info, I think, is best placed in the code comments, not in the commit message.

Comment (loli10K) [inline module/zfs/zfs_ioctl.c]: >would it make sense to set received == TRUE unconditionally in the user-space for the snapdev only ?

No, `received` serves a totally different purpose and we need both TRUE and FALSE, we cannot set `received == TRUE` unconditionally from userland for snapdev.

The intention is to call `zfs_prop_set_special()` for every property, not just `snapdev`. This seems only natural and is just a coincidence we haven't needed this until now, just because snapdev is the only "special" inheritable property: for this reason i don't like adding a comment that would specify the obvious case ("_call zfs_prop_set_special() for every property, not just "snapdev"_") because it could be confusing to anyone reading the code.

Now, if someone wanted to know why it was implemented this way, when this was changed, and maybe how it was working before, they can read the VCS history which explains our reason to make this change in the commit message.

File: tests/zfs-tests/tests/functional/zvol/zvol_misc/zvol_misc_snapdev.ksh [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+# 3.3 Check inheritance on multiple levels
+log_must zfs snapshot $SUBSNAP
+log_must zfs inherit snapdev $SUBZVOL
+log_must zfs set snapdev=hidden $VOLFS
+log_must zfs set snapdev=visible $TESTPOOL
+verify_inherited 'snapdev' 'hidden' $SUBZVOL $VOLFS
+block_device_wait
+check_missing $SUBSNAPDEV
+blockdev_exists $SNAPDEV
+
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/zvol/zvol_misc/zvol_misc_snapdev.ksh]: Yes, please do.  It wouldn't hurt to verify the expected block device state after each `zfs set` call too, even if we don't expect it to have changed.

File: module/zfs/zfs_ioctl.c (hunk 2 above)
Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: @loli10K 
my concern is: you intended to modify behavior for snapdev property only, yet the code changes affect behavior of all the properties when received is FALSE.
To my knowledge, everything but the snapdev property worked fine as far as inheritance went. Why do we need to change behavior for other properties ?

Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: One way to address this is to make the `zvol_set_prop_special()` conditional on `(receive == TRUE || property == snapdev)` or something of this nature. Then you could add comments describing why we need to call `zvol_set_prop_special()` for snapdev even when receive is FALSE.

File: module/zfs/zvol.c [hunk 4]
```diff
@@ -2239,18 +2236,31 @@ zvol_set_snapdev_sync_cb(dsl_pool_t *dp, dsl_dataset_t *ds, void *arg)
 }
 
 /*
- * Traverse all child snapshot datasets and apply snapdev appropriately.
+ * Traverse all child datasets and apply snapdev appropriately.
+ * We call dsl_prop_set_sync_impl() here to set the value only on the toplevel
```
Comment (bprotopopov) [inline module/zfs/zvol.c]: @loli10K, @behlendorf 

I think the snapdev property and how it interacts with the property inheritance can benefit from clarification at this time.

The original code has always propagated the `snapdev` settings for the top-level filesystem through all the children filesystems, as well as all their children zvols.
```
[root@centos-6 zfs-bprotopopov]# for z in $(zfs list -H | cut -f1); do zfs get -H snapdev $z; done
tpool	snapdev	hidden	default
tpool/top-fs	snapdev	visible	local
tpool/top-fs/sub-fs	snapdev	visible	local
tpool/top-fs/sub-fs/zvol0	snapdev	visible	local
tpool/top-fs/sub-fs/zvol1	snapdev	visible	local
tpool/top-fs/zvol0	snapdev	visible	local
tpool/top-fs/zvol1	snapdev	visible	local
tpool/zvol0	snapdev	hidden	local
[root@centos-6 zfs-bprotopopov]# zfs set snapdev=hidden tpool/top-fs
[root@centos-6 zfs-bprotopopov]# for z in $(zfs list -H | cut -f1); do zfs get -H snapdev $z; done
tpool	snapdev	hidden	default
tpool/top-fs	snapdev	hidden	local
tpool/top-fs/sub-fs	snapdev	hidden	local
tpool/top-fs/sub-fs/zvol0	snapdev	hidden	local
tpool/top-fs/sub-fs/zvol1	snapdev	hidden	local
tpool/top-fs/zvol0	snapdev	hidden	local
tpool/top-fs/zvol1	snapdev	hidden	local
tpool/zvol0	snapdev	hidden	local
[root@centos-6 zfs-bprotopopov]#
```
But I believe @loli10K  suggests that this is not quite correct, and that the propagation of the properties needs to be in line with the `zfs inherit` semantics. This seems reasonable to me.

As I understand it, `zfs inherit` command explicitly passes the values of the properties to a specific child or recursively to all the children, but it recurses in the user space (`zfs_for_each()`), so that kernel methods do not need to be concerned with this. If that's the case, then I believe there is an opportunity here to cleanup the code a bit: we should set the value for the dataset in question and then we should only be recursing over snapshots if the dataset is a zvol.

I don't think we need to recurse through the hierarchy and re-read and re-set snapdev for all the children. Also, before setting and recursing, let's check that the value to be set is different from the one that is already set. This can save quite a bit of I/O.

@loli10K has previously brought up a 'hypothetical' use case of changing the setting of `zvol_inhibit_dev` along with setting `snapdev` and suggested that 're-reading and re-setting' would make that use case work in a more intuitive way. I am not convinced this is an important use case to support cleanly (there is a simple workaround of toggling `snapdev` that also works). In my experience, setting zvol_inhibit_dev is a design-time decision, not a runtime one.

Thoughts ?

Comment (loli10K) [inline module/zfs/zvol.c]: > we should set the value for the dataset in question and then we should only be recursing over snapshots if the dataset is a zvol.

This would not deal with the following case:

```
root@debian-8-zfs:~# zfs get snapdev
NAME              PROPERTY  VALUE    SOURCE
testpool          snapdev   hidden   local                   <- zfs set snapdev=visible here
testpool/fs       snapdev   hidden   inherited from testpool
testpool/fs/zvol  snapdev   hidden   inherited from testpool <- snapshot is here
```

Overall, i think we're getting sidetracked here: yes, there can be performance optimizations, but this PR is just about fixing a couple of issues in snapdev inheritance.

File: module/zfs/zfs_ioctl.c (hunk 2 above)
Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: OK, let's clarify. You said: "...The behavior won't change: every "special" property (except snapdev) we have today is not inheritable."
But what I  was pointing out was that `zfs_set_prop_special()` is called in your code for all kinds of properties, including user properties, special, and non-special inheritable properties, which seems totally superficial and unnecessary. Incidental to that, there are nvlists created and freed, arguments prepared for `zfs_set_prop_special()` only to have it return -1 and generate a bunch of heat and no other effect.
To summarize, I don't like the way you have restructured the code paths. I find it much harder to understand what's going on, and the hole bunch of no-op processing occurring for no reason whatsoever is something I that prefer to avoid.

File: module/zfs/zvol.c (hunk 4 above)
Comment (bprotopopov) [inline module/zfs/zvol.c]: What is the use case you are referring to (please give the command sequence) ?
Please provide more detail.

File: module/zfs/zfs_ioctl.c (hunk 2 above)
Comment (loli10K) [inline module/zfs/zfs_ioctl.c]: > zfs_set_prop_special() is called in your code for all kinds of properties, including user properties, special, and non-special inheritable properties, which seems totally superficial and unnecessary. Incidental to that, there are nvlists created and freed, arguments prepared for zfs_set_prop_special() only to have it return -1 and generate a bunch of heat and no other effect.

> To summarize, I don't like the way you have restructured the code paths. I find it much harder to understand what's going on, and the hole bunch of no-op processing occurring for no reason whatsoever is something I that prefer to avoid.

Can you propose a prototype (even pseudocode is fine) that would allow us to avoid unnecessary processing (no-ops, nvlist alloc/free, calling `zfs_set_prop_special()` for non-"special" properties...) but still retain the same functionality? Thanks

Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: @loli10K 

yes, a simple way to control the overheads is to introduce a `PROP_SPECIAL` flag to the `zfs_prop_table`,  that could be checked for (`boolean_t zfs_prop_is_special(zfs_prop_t prop)`, similar to `zfs_prop_inheritable()`) before calling `zfs_prop_set_special()` and going through the incidental overhead.

That said, I am looking through the code that uses `zfs_prop_set_special()`, and I see that what you do with these code changes is not out of line with the current uses of `zfs_prop_set_special()`, and the incidental inefficiencies are not addressed elsewhere either. Also, this is probably not performance/resource critical, as property setting is not a frequent/concurrent operation anyway. So, if you don't feel like doing much about this, that's fine with me.

More to the subject matter of you PR, I have experimented a bit with setting inheritable properties across namespace hierarchies that involve zvols, and I found that the way the source of `snapdev` is set is not consistent with other inheritable properties. It appears that simple setting of an inheritable property (e.g. `dedup`) on a top filesystem automagically inherits that value for all the descendants (filesystems and zvols), and the source is marked as such (inherited from top filesystem).  I also found that this behavior changes if one explicitly sets or inherits from some filesystem in the hierarchy - the subsequent changes of top-level property no longer affect the subtree that was inherited explicitly, and the source of inheritance is properly tracked (from specific filesystem and form the top level).

In case of `snapdev` however, setting the top-level filesystem property results in the values propagated but the source  is 'local', not 'inherited from ...'. Also, regardless of explicit setting of `snapdev` in the middle of the hierarchy, the changes to top-level fs are always propagated down the hierarchy, which is not quite how this works for `dedup`.

All of this is easy to see on a toy pool with a few datasets. If it helps, I can attach a terminal session log that illustrates the points made above.

To summarize, as I understand it, there are two issues addressed in this PR:
1) NULL pointer dereference in `zvol_free()`
2) tracking of snapdev property source that is inconsistent with other properties

These two issues are quite separate, and the second does not seem to be fully addressed with your changes. Therefore, I recommend that you split this PR in two, and work separately on addressing the inheritance source/value propagation inconsistencies for `snapdev`.

Comment (loli10K) [inline module/zfs/zfs_ioctl.c]: > In case of snapdev however, setting the top-level filesystem property results in the values propagated but the source is 'local', not 'inherited from ...'. Also, regardless of explicit setting of snapdev in the middle of the hierarchy, the changes to top-level fs are always propagated down the hierarchy, which is not quite how this works for dedup.

> If it helps, I can attach a terminal session log that illustrates the points made above.

Please do, my testing is limited but my changes seem to be working as intended:

```
root@zfstest:~# # ensure we're running the correct zfs version
root@zfstest:~# cat /sys/module/zfs/version
0.7.0-rc4_39_g5fc10596e
root@zfstest:~# #
root@zfstest:~# # setup
root@zfstest:~# POOLNAME='testpool'
root@zfstest:~# TMPDIR='/var/tmp'
root@zfstest:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@zfstest:~# zpool destroy $POOLNAME
root@zfstest:~# rm -f $TMPDIR/zpool.dat
root@zfstest:~# fallocate -l 128m $TMPDIR/zpool.dat
root@zfstest:~# zpool create -O mountpoint=none $POOLNAME $TMPDIR/zpool.dat
root@zfstest:~# zfs create $POOLNAME/fs1
root@zfstest:~# zfs create $POOLNAME/fs2
root@zfstest:~# zfs create -V 1M -s $POOLNAME/fs1/zvol11
root@zfstest:~# zfs create -V 1M -s $POOLNAME/fs1/zvol12
root@zfstest:~# zfs create -V 1M -s $POOLNAME/fs2/zvol21
root@zfstest:~# #
root@zfstest:~# # everything is default
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   hidden   -         default
testpool/fs1         snapdev   hidden   -         default
testpool/fs1/zvol11  snapdev   hidden   -         default
testpool/fs1/zvol12  snapdev   hidden   -         default
testpool/fs2         snapdev   hidden   -         default
testpool/fs2/zvol21  snapdev   hidden   -         default
root@zfstest:~# #
root@zfstest:~# # every dataset inherit from top-level
root@zfstest:~# zfs set snapdev=visible $POOLNAME
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   visible  -         local
testpool/fs1         snapdev   visible  -         inherited from testpool
testpool/fs1/zvol11  snapdev   visible  -         inherited from testpool
testpool/fs1/zvol12  snapdev   visible  -         inherited from testpool
testpool/fs2         snapdev   visible  -         inherited from testpool
testpool/fs2/zvol21  snapdev   visible  -         inherited from testpool
root@zfstest:~# #
root@zfstest:~# # fs1 is 'local', its children inherit. everything else is same as before
root@zfstest:~# zfs set snapdev=hidden $POOLNAME/fs1
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   visible  -         local
testpool/fs1         snapdev   hidden   -         local
testpool/fs1/zvol11  snapdev   hidden   -         inherited from testpool/fs1
testpool/fs1/zvol12  snapdev   hidden   -         inherited from testpool/fs1
testpool/fs2         snapdev   visible  -         inherited from testpool
testpool/fs2/zvol21  snapdev   visible  -         inherited from testpool
root@zfstest:~# #
root@zfstest:~# # fs1 will now inherit from the top-level, again
root@zfstest:~# zfs inherit snapdev $POOLNAME/fs1
root@zfstest:~# zfs get -o all snapdev -r $POOLNAME
NAME                 PROPERTY  VALUE    RECEIVED  SOURCE
testpool             snapdev   visible  -         local
testpool/fs1         snapdev   visible  -         inherited from testpool
testpool/fs1/zvol11  snapdev   visible  -         inherited from testpool
testpool/fs1/zvol12  snapdev   visible  -         inherited from testpool
testpool/fs2         snapdev   visible  -         inherited from testpool
testpool/fs2/zvol21  snapdev   visible  -         inherited from testpool
```

Comment (bprotopopov) [inline module/zfs/zfs_ioctl.c]: @loli10K 

I have rebuilt/re-installed/re-tested, and no longer see the problems I have described above. There must have been some inconsistency in my environment. 

My testing also shows that snapdev behaves consistently with dedup, for instance. 
Please disregard my earlier comment.

File: module/zfs/zvol.c (hunk 4 above)
Comment (bprotopopov) [inline module/zfs/zvol.c]: @loli10K 
still looking for more detail on this.

Comment (loli10K) [inline module/zfs/zvol.c]: Given the following situation
```
NAME                    PROPERTY  VALUE    SOURCE
testpool                snapdev   hidden   default
testpool/fs             snapdev   hidden   default
testpool/fs/zvol        snapdev   hidden   default
testpool/fs/zvol@snap1  snapdev   -        -
```
> we should set the value for the dataset in question

which translates to `zfs set snapdev=visible testpool`

> and then **we should only be recursing over snapshots if the dataset is a zvol**

which means we should do nothing since `testpool` is not a zvol: we won't create `testpool/fs/zvol@snap1` device node, even if now we have `snapdev=visible` on the zvol:

```
NAME                    PROPERTY  VALUE    SOURCE
testpool                snapdev   visible  local
testpool/fs             snapdev   visible  inherited from testpool
testpool/fs/zvol        snapdev   visible  inherited from testpool
testpool/fs/zvol@snap1  snapdev   -        -
```

Comment (bprotopopov) [inline module/zfs/zvol.c]: Yes, that's a good point, I guess I forgot how this actually worked.

**Comment by behlendorf** [comment]
@loli10K @bprotopopov thanks for hashing this out.  The change LGTM too, I'll get this merged.
